### PR TITLE
feat: fallback MGetCache to MGet

### PR DESCRIPTION
**Pull request by SoulPancake**
Solves for #567

### CODE REVIEW

File: helper.go [hunk 1]
```diff
@@ -54,6 +57,22 @@ func MGetCache(client Client, ctx context.Context, ttl time.Duration, keys []str
 	return doMultiCache(client, ctx, cmds.s, keys)
 }
 
+func isCacheDisabled(client Client) bool {
+	switch c := client.(type) {
+	case *singleClient:
+		// what do we do here ? @rueian
```
Comment (SoulPancake) [inline helper.go:63]: Is this the right approach to this? @rueian 
Can you suggest?

Comment (rueian) [reply helper.go:63]: We may copy the `DisableCache` to a new bool field in the `singleClient` and then check the field here.

File: helper.go [hunk 2]
```diff
@@ -54,6 +57,22 @@ func MGetCache(client Client, ctx context.Context, ttl time.Duration, keys []str
...
+		return false
+	case *sentinelClient:
+		if c.mOpt != nil && c.mOpt.DisableCache {
+			return true
+		}
+		if c.sOpt != nil && c.sOpt.DisableCache {
+			return true
+		}
+	}
```
Comment (rueian) [inline helper.go:72]: We should also do something similar for the `clusterClient`.

File: sentinel.go [hunk 3]
```diff
@@ -177,7 +177,11 @@ func (c *sentinelClient) Dedicate() (DedicatedClient, func()) {
 
 func (c *sentinelClient) Nodes() map[string]Client {
 	conn := c.mConn.Load().(conn)
-	return map[string]Client{conn.Addr(): newSingleClientWithConn(conn, c.cmd, c.retry)}
+	disableCache := false
+	if (c.mOpt != nil && c.mOpt.DisableCache) || (c.sOpt != nil && c.sOpt.DisableCache) {
```
Comment (rueian) [inline sentinel.go:181]: We don't need to consider `c.sOpt` here. That is basically a copy of `c.mOpt`.

File: helper.go [hunk 4]
```diff
@@ -54,6 +57,21 @@ func MGetCache(client Client, ctx context.Context, ttl time.Duration, keys []str
 	return doMultiCache(client, ctx, cmds.s, keys)
 }
 
+func isCacheDisabled(client Client) bool {
+	switch c := client.(type) {
+	case *singleClient:
+		if c.DisableCache {
+			return true
+		}
+		return false
```
Comment (rueian) [inline helper.go:66]:
```suggestion
		return c.DisableCache
```

File: helper.go [hunk 5]
```diff
@@ -54,6 +57,21 @@ func MGetCache(client Client, ctx context.Context, ttl time.Duration, keys []str
...
+	case *sentinelClient:
+		return (c.mOpt != nil && c.mOpt.DisableCache) || (c.sOpt != nil && c.sOpt.DisableCache)
```
Comment (rueian) [inline helper.go:68]:
```suggestion
		return (c.mOpt != nil && c.mOpt.DisableCache)
```

File: helper.go [hunk 6]
```diff
@@ -54,6 +57,21 @@ func MGetCache(client Client, ctx context.Context, ttl time.Duration, keys []str
...
+	case *clusterClient:
+		return (c.opt != nil && c.opt.DisableCache) || (c.rOpt != nil && c.rOpt.DisableCache)
```
Comment (rueian) [inline helper.go:70]:
```suggestion
		return (c.opt != nil && c.opt.DisableCache)
```

File: helper.go [hunk 7]
```diff
@@ -46,6 +46,9 @@ func MGetCache(client Client, ctx context.Context, ttl time.Duration, keys []str
 	if len(keys) == 0 {
 		return make(map[string]RedisMessage), nil
 	}
+	if isCacheDisabled(client) {
+		return MGet(client, ctx, keys)
```
Comment (rueian) [inline helper.go:50]: We need tests on this.

Comment (SoulPancake) [reply helper.go:50]: Will update the tests and cover these conditions

File: helper_test.go [hunk 8]
```diff
@@ -108,6 +125,35 @@ func TestMGetCache(t *testing.T) {
 			}
 		})
 	})
+	// TODO: @SoulPancake I have some doubts about this @rueian
```
Comment (SoulPancake) [inline helper_test.go:128]: I have some doubts here, Can you suggest how we can cover the sentinelClient scenario? @rueian

Comment (rueian) [reply helper_test.go:128]: Here is an example:  https://github.com/redis/rueidis/blob/63951b5b27a90738fe251a48305f65954df3fec9/sentinel_test.go#L664

File: helper.go (hunk 7 above)
Comment (SoulPancake) [reply helper.go:50]: I have covered that case, can you please take a look again? @rueian

File: helper_test.go [hunk 9]
```diff
@@ -18,6 +18,23 @@ func TestMGetCache(t *testing.T) {
 		if err != nil {
 			t.Fatalf("unexpected err %v", err)
 		}
+		disabledCacheClient, err := newSingleClient(&ClientOption{InitAddress: []string{""}, DisableCache: true}, m, func(dst string, opt *ClientOption) conn {
```
Comment (rueian) [inline helper_test.go:21]: Hi @SoulPancake, we still need a test for the `newClusterClient`.

File: helper_test.go [hunk 10]
```diff
@@ -55,13 +72,49 @@ func TestMGetCache(t *testing.T) {
...
+				return &redisresults{s: []RedisResult{newErrResult(context.Canceled), newErrResult(context.Canceled)}}
+			},
 		}
 		client, err := newClusterClient(&ClientOption{InitAddress: []string{":0"}}, func(dst string, opt *ClientOption) conn {
 			return m
 		})
 		if err != nil {
 			t.Fatalf("unexpected err %v", err)
 		}
+		disabledCacheClient, err := newClusterClient(&ClientOption{InitAddress: []string{":0"}, DisableCache: true}, func(dst string, opt *ClientOption) conn {
```
Comment (SoulPancake) [inline helper_test.go:85]: I have some doubts about this, and it is timing out, Can you take a look? @rueian

Comment (rueian) [reply helper_test.go:85]: Hi @SoulPancake, when the `DisableCache` is set, you should mock the m.DoMultiFn instead of the m.DoMultiCacheFn.

Comment (SoulPancake) [reply helper_test.go:85]: Done, Can you please review @rueian

File: helper_test.go [hunk 11]
```diff
@@ -55,13 +72,49 @@ func TestMGetCache(t *testing.T) {
 			DoFn: func(cmd Completed) RedisResult {
 				return slotsResp
 			},
+			DoMultiFn: func(cmd ...Completed) *redisresults {
+				return &redisresults{s: []RedisResult{newErrResult(context.Canceled), newErrResult(context.Canceled)}}
```
[the remark this answers is not in the document]
Comment (SoulPancake) [reply helper_test.go]: removed

**Review by rueian — APPROVED** [review]

**Comment by rueian** [comment]
Thank you @SoulPancake!

**Pull request merged**